Makefile

```diff
@@ -0,0 +1,64 @@
+
+# Image URL to use all building/pushing image targets
+IMG ?= controller:latest
+# Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
+CRD_OPTIONS ?= "crd:trivialVersions=true"
+
+all: manager
+
+# Run tests
+test: generate fmt vet manifests
+	go test ./api/... ./controllers/... -coverprofile cover.out
+
+# Build manager binary
+manager: generate fmt vet
+	go build -o bin/manager main.go
+
+# Run against the configured Kubernetes cluster in ~/.kube/config
+run: generate fmt vet
+	go run ./main.go
+
+# Install CRDs into a cluster
+install: manifests
+	kubectl apply -f config/crd/bases
+
+# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
+deploy: manifests
+	kubectl apply -f config/crd/bases
+	kustomize build config/default | kubectl apply -f -
+
+# Generate manifests e.g. CRD, RBAC etc.
+manifests: controller-gen
+	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+
+# Run go fmt against code
+fmt:
+	go fmt ./...
+
+# Run go vet against code
+vet:
+	go vet ./...
+
+# Generate code
+generate: controller-gen
+	$(CONTROLLER_GEN) object:headerFile=./hack/boilerplate.go.txt paths=./api/...
+
+# Build the docker image
+docker-build: test
+	docker build . -t ${IMG}
+	@echo "updating kustomize image patch file for manager resource"
+	sed -i'' -e 's@image: .*@image: '"${IMG}"'@' ./config/default/manager_image_patch.yaml
+
+# Push the docker image
+docker-push:
+	docker push ${IMG}
+
+# find or download controller-gen
+# download controller-gen if necessary
+controller-gen:
+ifeq (, $(shell which controller-gen))
+	go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.0-beta.2
+CONTROLLER_GEN=$(shell go env GOPATH)/bin/controller-gen
+else
+CONTROLLER_GEN=$(shell which controller-gen)
```

Comment:
I'd generally advise building utilities from vendor and putting them in a bin/ folder

Reply:
vendor is being phased out in favor of a go proxy which we can leverage using a script like this instead of a direct go get:

```
#!/usr/bin/env bash
VERSION=${VERSION:-0.2.0-beta.2}

# get out of the module
cd /tmp
GO111MODULE=on go get "sigs.k8s.io/controller-tools/cmd/controller-gen@v${VERSION}"
cd -
```

Here is my thinking that led me down this path:

* we can't use `go get` within a go module (as you pointed out, it modifies go.mod/go.sum)
* controller-runtime doesn't release binaries so we must build from source if we don't have `controller-gen` installed (this makes using go get VERY appealing)
* vendor is being phased out in favor a proxy (don't vendor the source to build from vendor)
* outside of the GOPATH we can't use the version specification (must be outside of any module but still use go get with module mode)

which is how this script ended up with getting outside of the module but still using go get (in module mode) to install the dependency